### PR TITLE
 API - User features

### PR DESCRIPTION
**Pull request by zenjutahi**
#### What this PR does
API creates a user account
API logs in a user to their account
API logs out a user

API registers user business
API retrieves all registred businesses

#### Tasks completed
- User model created and implemented
- Created Tests for user authentication functions
- Created user authentication functions; 1. register 2. login 3. logout
- Run tests for the user authentication functions 
- Debugged the tests and functions and made sure they run

#### How this should be manually tested
Using postman by giving it the routes for endpoints

#### Stories covered
[ Feature #155688213]
[ Feature #155688257]
[ Feature #155690994]
[ Feature ##155502275]

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,102 @@
...
+Give an example
+```
+
+## Deployment
+
+Add additional notes about how to deploy this on a live system
+
+## Built With
+
+* [Dropwizard](http://www.dropwizard.io/1.0.2/docs/) - The web framework used
```
Comment (georgreen) [inline README.md:75]: Change this to the framework you are using, Flask

File: Restfull API/api/app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+from flask import Flask, Markup
+#local imports
+from config import app_config
+# Initialize the app
+
+def create_app(config_name):
```
Comment (georgreen) [inline Restfull API/api/app/__init__.py:6]: What does this function do?

File: Restfull API/api/app/auth/__init__.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from flask import Blueprint
+
+auth = Blueprint('auth', __name__)
```
Comment (georgreen) [inline Restfull API/api/app/auth/__init__.py:3]: What is a blueprint?

File: README.md (hunk 1 above)
Comment (zenjutahi) [reply README.md:75]: yup i will do that

File: Restfull API/api/app/models.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+from flask import session, request
```
Comment (georgreen) [inline Restfull API/api/app/models.py:1]: NOTE: You are building an API, why do you have sessions? API should not have session data.

File: Restfull API/api/app/__init__.py (hunk 2 above)
Comment (zenjutahi) [reply Restfull API/api/app/__init__.py:6]: well that is the code that launches my app

File: Restfull API/api/instance/config.py [hunk 5]
```diff
@@ -0,0 +1 @@
+SECRET_KEY = 'p9Bv<3Eid9%$i01'
```
Comment (georgreen) [inline Restfull API/api/instance/config.py]: You can move this into the Config class that holds the common configurations for the app.

File: Restfull API/api/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+itsdangerous==0.24
+Jinja2==2.10
+lazy-object-proxy==1.3.1
+MarkupSafe==1.0
+mccabe==0.6.1
+pylint==1.8.2
+six==1.11.0
+Werkzeug==0.14.1
+wrapt==1.10.11
+WTForms==2.1
```
Comment (georgreen) [inline Restfull API/api/requirements.txt:16]: Why do you have WTFForms in your requirements.txt? Are you using this?

File: Restfull API/api/tests/test_auth.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
+import unittest
+import os
+import json
+import sys
+import inspect
+
+currentdir = os.path.dirname(os.path.abspath(
+  inspect.getfile(inspect.currentframe())))
+parentdir = os.path.dirname(currentdir)
+sys.path.insert(0, parentdir)
```
Comment (georgreen) [inline Restfull API/api/tests/test_auth.py:10]: What is line 7 to line 10 for?

File: Restfull API/api/app/auth/__init__.py (hunk 3 above)
Comment (zenjutahi) [reply Restfull API/api/app/auth/__init__.py:3]: A blueprint is more like packages for the different sections of my app. Am using it to separate the different features or sections of my app example, authentication is one section, business will be another...so on so forth.

File: Restfull API/api/tests/test_auth.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        self.assertIn("valid", response_msg["message"])  
+
+
+    def test_user_login_with_wrong_password_or_username(self):
+        """ Test API can check if user used wrong password"""
+        response = self.app.post("/auth/login",
+                        data=json.dumps(dict(email="jeff@gmail.com",password="jeffpassxxx")),
+                                         content_type="application/json")
+
+        self.assertEqual(response.status_code, 404)
```
Comment (georgreen) [inline Restfull API/api/tests/test_auth.py]: Your status code is not correct, when someone tries to login with wrong password the http status code is not 404.

File: Restfull API/api/app/models.py (hunk 4 above)
Comment (zenjutahi) [reply Restfull API/api/app/models.py:1]: Note: i should look at jwt

File: Restfull API/api/tests/test_auth.py (hunk 7 above)
Comment (zenjutahi) [reply Restfull API/api/tests/test_auth.py:10]: I had problems importing some functions from my other modules, so I used that code to give my text file the intended path

File: Restfull API/api/requirements.txt (hunk 6 above)
Comment (zenjutahi) [reply Restfull API/api/requirements.txt:16]: At the moment no.I initially thought I'd use forms

File: Restfull API/api/tests/test_auth.py (hunk 8 above)
Comment (zenjutahi) [reply Restfull API/api/tests/test_auth.py]: I will definitely edit that thanks

**Pull request merged**